packages/@react-stately/table/src/useTableColumnResizeState.ts

```diff
@@ -0,0 +1,181 @@
+  /** Callback that is invoked during the entirety of the resize event. */
+  onColumnResize?: (affectedColumnWidths: { key: Key, width: number }[]) => void,
+  /** Callback that is invoked when the resize event is ended. */
+  onColumnResizeEnd?: (affectedColumnWidths: { key: Key, width: number }[]) => void
+}
+
+export default function useTableColumnResizeState<T>(props: ColumnResizeStateProps<T>): ColumnResizeState<T> {
+
+  const {columns, getDefaultWidth} = props;
+  const columnsRef = useRef<GridNode<T>[]>();
+  const tableWidth = useRef<number>(null);
+  const isResizing = useRef<boolean>(null);
+  const startResizeContentWidth = useRef<number>();
+
+  const [columnWidths, setColumnWidths] = useState<Map<Key, number>>(new Map(columns.map(col => [col.key, 0])));
+  const columnWidthsRef = useRef<Map<Key, number>>(columnWidths);
+  const [resizedColumns, setResizedColumns] = useState<Set<Key>>(new Set());
+  const resizedColumnsRef = useRef<Set<Key>>(resizedColumns);
+
+  let getRealColumnWidth = useCallback((column: GridNode<T>) => {
+    let props = column.props as ColumnProps<T>;
+    return resizedColumns?.has(column.key) ? columnWidthsRef.current.get(column.key) : props.width ?? props.defaultWidth ?? getDefaultWidth?.(column.props) ?? '1fr';
+  }, [getDefaultWidth, resizedColumns]);
+
+  let getStaticAndDynamicColumns = useCallback((columns: GridNode<T>[]) : { staticColumns: GridNode<T>[], dynamicColumns: GridNode<T>[] } => columns.reduce((acc, column) => {
+    let width = getRealColumnWidth(column);
+    return isStatic(width) ? {...acc, staticColumns: [...acc.staticColumns, column]} : {...acc, dynamicColumns: [...acc.dynamicColumns, column]}; 
+  }, {staticColumns: [], dynamicColumns: []}), [getRealColumnWidth]);
+
+  let buildColumnWidths = useCallback((affectedColumns: GridNode<T>[], availableSpace: number): Map<Key, number> => {
+    const widths = new Map<Key, number>();
+    let remainingSpace = availableSpace;
+
+    const {staticColumns, dynamicColumns} = getStaticAndDynamicColumns(affectedColumns);
+
+    staticColumns.forEach(column => {
+      let width = getRealColumnWidth(column);
+      let w = parseStaticWidth(width, tableWidth.current);
+      widths.set(column.key, w);
+      remainingSpace -= w;
+    });
+
+    // dynamic columns
+    if (dynamicColumns.length > 0) {
+      const newColumnWidths = getDynamicColumnWidths(dynamicColumns, remainingSpace, tableWidth.current);
+      for (let column of newColumnWidths) {
+        widths.set(column.key, column.calculatedWidth);
+      }
+    }
+
+    return widths;
+  }, [getStaticAndDynamicColumns, getRealColumnWidth]);
+
+  // if the columns change, need to rebuild widths.
+  useLayoutEffect(() => {
+    columnsRef.current = columns;
+    const widths = buildColumnWidths(columns, tableWidth.current);
+    setColumnWidthsForRef(widths);
+  }, [columns, buildColumnWidths]);
+
+  function setColumnWidthsForRef(newWidths: Map<Key, number>) {
```

Comment:
move this function up nearer to where the state and ref are defined so it's obvious how to set them both at the same time